lib/pbench/server/api/resources/users_api.py

```diff
@@ -0,0 +1,541 @@
+        except Exception:
+            self.logger.exception("Exception occurred during the getUser")
+            abort(500, message="INTERNAL ERROR")
+
+        # TODO: Check if the user has the right privileges
+        if verified or user.is_admin():
+            response_object = {
+                "status": "success",
+                "data": {
+                    "email": user.email,
+                    "first_name": user.first_name,
+                    "last_name": user.last_name,
+                    "registered_on": user.registered_on,
+                },
+            }
+            return make_response(jsonify(response_object), 200)
+        else:
+            self.logger.warning(
+                "Username retrieved from the auth token {} is different from the username provided",
+                self.auth.token_auth.current_user().id,
+            )
+            abort(403, message="Authentication token does not belong to the user")
+
+    @Auth.token_auth.login_required(f=Auth().verify_auth())
+    def put(self, username):
+        """
+        PUT request for updating user data.
+        This requires a Pbench auth token in the header field
+
+        This requires a JSON data with required user registration fields that needs an update
+        Example Json:
+        {
+            "first_name": "new_name",
+            "password": "password",
+            ...
+        }
+
+        Required headers include
+
+            Content-Type:   application/json
+            Accept:         application/json
+            Authorization:  Bearer Pbench_auth_token (user received upon login)
+
+        :return: JSON Payload
+            response_object = {
+                    "status": "success"/"fail",
+                    "message": "Successfully registered."/"failure message",
+                    "data": {
+                        "username": <username>,
+                        "first_name": <firstName>,
+                        "last_name": <lastName>,
+                        "registered_on": registered_on,
+                    }
+                }
+        """
+        post_data = request.get_json()
+        if not post_data:
+            self.logger.warning("Invalid json object: {}", request.url)
+            abort(400, message="Invalid json object in request")
+
+        try:
+            user, verified = self.auth.verify_user(username)
+        except Exception:
+            self.logger.exception("Exception occurred while verifying the user")
+            abort(500, message="INTERNAL ERROR")
+
+        # TODO: Check if the user has the right privileges
+        if verified:
+            try:
+                # Log if the user payload contain fields that are either non-updatabale or
+                # are not present in the user db.
+                non_updatable = list(
+                    set(post_data.keys()) - set(User.__table__.columns.keys())
+                )
+                if "registered_on" in post_data.keys():
```

Comment:
This conditional check is just for the logging purpose, the decision about which fields to update happens within the user model `users.py` file. Also the `User()` is a sqlalchemy class and a corresponding object `user` is always queried from our users database therefore creating a `frozenset` inside User class would not help since the user object is a sql query and not a python class object.